tests/test_zernike.py

```diff
@@ -349,6 +348,63 @@ def test_fit():
         np.testing.assert_allclose(resids2, 0, atol=1e-14)
 
 
+@timer
+def test_gradient():
+    # Start with a few that just quote the literature, e.g., Stephenson (2014).
+
+    Z11 = galsim.zernike.Zernike([0]*11+[1])
+
+    x = np.linspace(-1, 1, 100)
+    x, y = np.meshgrid(x, x)
+
+    def Z11_grad(x, y):
+        r2 = x**2 + y**2
+        gradx = 12*np.sqrt(5)*x*(2*r2-1)
+        grady = 12*np.sqrt(5)*y*(2*r2-1)
+        return gradx, grady
+
+    # import matplotlib.pyplot as plt
+    # fig, axes = plt.subplots(ncols=3, figsize=(12, 3))
+    # scat0 = axes[0].scatter(x, y, c=Z11.evalCartesianGrad(x, y)[0])
+    # fig.colorbar(scat0, ax=axes[0])
+    # scat1 = axes[1].scatter(x, y, c=Z11_grad(x, y)[0])
+    # fig.colorbar(scat1, ax=axes[1])
+    # scat2 = axes[2].scatter(x, y, c=Z11.evalCartesianGrad(x, y)[0] - Z11_grad(x, y)[0])
+    # fig.colorbar(scat2, ax=axes[2])
+    # plt.show()
+
+    np.testing.assert_allclose(Z11.evalCartesianGrad(x, y), Z11_grad(x, y), rtol=0, atol=1e-12)
+
+    Z28 = galsim.zernike.Zernike([0]*28+[1])
+
+    def Z28_grad(x, y):
+        gradx = 6*np.sqrt(14)*x*(x**4-10*x**2*y**2+5*y**4)
+        grady = -6*np.sqrt(14)*y*(5*x**4-10*x**2*y**2+y**4)
+        return gradx, grady
+
+    np.testing.assert_allclose(Z28.evalCartesianGrad(x, y), Z28_grad(x, y), rtol=0, atol=1e-12)
+
+    # Now try some finite differences on a broader set of input
+
+    def finite_difference_gradient(Z, x, y):
+        dh = 1e-5
+        return ((Z.evalCartesian(x+dh, y)-Z.evalCartesian(x-dh, y))/(2*dh),
+                (Z.evalCartesian(x, y+dh)-Z.evalCartesian(x, y-dh))/(2*dh))
+
+    u = galsim.UniformDeviate(1234)
+
+    for j in range(25):
+        nj = 1+int(u()*55)
+        R_inner = 0.2+0.6*u()
+        R_outer = R_inner + 0.2+0.6*u()
+        Z = galsim.zernike.Zernike([0]+[u() for _ in range(nj)], R_inner=R_inner, R_outer=R_outer)
```

Comment:
This could also use some inline commenting.  I think I understand that all this is just to generate 25 random zernike arrays with random inner and outer radii, so it constitutes a fairly comprehensive test of the correctness of the derivative.  But it would be helpful to just explain that a bit in comments.